firedrake/eigensolver.py

```diff
@@ -0,0 +1,192 @@
+    eigenproblem having a nullspace spanned by the basis functions with support
+    on the boundary. To facilitate solution, this is shifted by the specified
+    amount. It is the user's responsibility to ensure that the shift is not
+    close to an actual eigenvalue of the system.
+    """
+    def __init__(self, A, M=None, bcs=None, bc_shift=666.0):
+        if not SLEPc:
+            raise ImportError(
+                "Unable to import SLEPc, eigenvalue computation not possible "
+                "(try firedrake-update --slepc)"
+            )
+
+        self.A = A  # LHS
+        args = A.arguments()
+        v, u = args
+        if M:
+            self.M = M
+        else:
+            from ufl import inner, dx
+            self.M = inner(u, v) * dx
+        self.output_space = u.function_space()
+        self.bcs = bcs
+        self.bc_shift = bc_shift
+
+    def dirichlet_bcs(self):
+        """Return an iterator over the Dirichlet boundary conditions."""
+        for bc in self.bcs:
+            yield from bc.dirichlet_bcs()
+
+    @utils.cached_property
+    def dm(self):  # cargo cult
+        r"""Return the dm associated with the output space."""
+        return self.output_space.dm
+
+
+class LinearEigensolver(OptionsManager):
+    r"""Solve a LinearEigenproblem.
+
+    Parameters
+    ----------
+    problem : LinearEigenproblem
+        The eigenproblem to solve.
+    n_evals : int
+        The number of eigenvalues to compute.
+    options_prefix : str
+        The options prefix to use for the eigensolver.
+    solver_parameters : dict
+        PETSc options for the eigenvalue problem.
+    ncv : int
+        Maximum dimension of the subspace to be used by the solver. See
+        `SLEPc.EPS.setDimensions`.
+    mpd : int
+        Maximum dimension allowed for the projected problem. See
+        `SLEPc.EPS.setDimensions`.
+
+    Notes
+    -----
+
+    Users will typically wish to set solver parameters specifying the symmetry
+    of the eigenproblem and which eigenvalues to search for first.
+
+    The former is set using the options available for `EPSSetProblemType
+    <https://slepc.upv.es/documentation/current/docs/manualpages/EPS/EPSSetProblemType.html>`__.
+
+    For example if the bilinear form is symmetric (Hermitian in complex mode),
+    one would add this entry to `solver_options`::
+
+        "eps_gen_hermitian": None
+
+    As always when specifying PETSc options, `None` indicates that the option
+    in question is a flag and hence doesn't take an argument.
+
+    The eigenvalues to search for first are specified using the options for
+    `EPSSetWhichEigenPairs <https://slepc.upv.es/documentation/current/docs/manualpages/EPS/EPSSetWhichEigenpairs.html>`__.
+
+    For example, to look for the eigenvalues with largest real part, one
+    would add this entry to `solver_options`::
+
+        "eps_largest_real": None
+    """
+
+    DEFAULT_EPS_PARAMETERS = {"st_pc_factor_shift_type": "NONZERO",
```

Comment:
`"st_pc_factor_shift_type": "NONZERO" should only be used in case we have prior knowledge that the mass matrix has a zero pivot, no?